<a href="https://colab.research.google.com/github/Theivaprakasham/BIS/blob/master/UCI%20HAR%20Dataset%20_%20Machine%20Learning%20Algorithms%20comparision.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**M.Tech Data Science (2020-22)**

**20DS614 – Machine Learning for Signal Processing and Pattern Classification**

**Quiz-4**

**Date: 08.06.2021.**

**Time: 11.00 A.M. to 4:00 P.M.**

**Marks: 50 marks.**

***Submitted by: Theivaprakasham H (CB.EN.P2CEN20026)***

**Answer the question:**

**1.  For the given dataset, give the dataset description in the perspective of multi-class classification using Machine Learning. (5 marks).**

In [ ]:
!nvidia-smi

Tue Jun  8 16:32:23 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.27       Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   43C    P0    27W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
!wget "https://dl.dropboxusercontent.com/s/5eglqi59a9trk0r/Dataset.zip"

--2021-06-08 16:34:08--  https://dl.dropboxusercontent.com/s/5eglqi59a9trk0r/Dataset.zip
Resolving dl.dropboxusercontent.com (dl.dropboxusercontent.com)... 162.125.80.15, 2620:100:6016:15::a27d:10f
Connecting to dl.dropboxusercontent.com (dl.dropboxusercontent.com)|162.125.80.15|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 111368069 (106M) [application/zip]
Saving to: ‘Dataset.zip’

Dataset.zip         100%[===================>] 106.21M  24.2MB/s    in 5.4s    

2021-06-08 16:34:15 (19.8 MB/s) - ‘Dataset.zip’ saved [111368069/111368069]



In [ ]:
!unzip "/content/Dataset.zip"

Streaming output truncated to the last 5000 lines.
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__5_4515.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__4_6922.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__12_9564.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__1_3955.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__7_5121.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__8_3771.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__9_6938.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__6_2590.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__9_1592.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__2_4695.jpeg  
  inflating: Dataset/__MACOSX/Blood_Cell_Images/images/TRAIN/MONOCYTE/.__11_27

In [ ]:
!wget "https://storage.googleapis.com/kaggle-data-sets/7415/10564/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210608%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210608T163726Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=9b8bb3b3e0daebd7132e73f4c4e73243e9c1747c81b5dcbb675268e3e196f11f0eb77a4d96ae1e2effa2ae0aaecf71ec4ffc30664dc08a2e4df54e51dc4f386010550912f80266e7872eb309c2d78242448f897a62d6667d3ff358f30c65db3538de8f6ef1b87d505ce1447ec74e93b1d7177272ae04f3e4f9bac9d7a73f72a1f12161056d3cd1bded96c05b74e481c5139a9361be7a7576e4bef3205d17c148e198f3d46637d813ac694d3471ec70a205504de85fd8cb9e9580b6d51caa2e2c4f2f6760eb35c91d52cfe8197c4aa7291eab222e8ec0bda877f5f74ab1f96d8f5cb8fd336d91b1924376e76a39ff6d19c22b31297d65186712cbf0f7c113a4ee"

The name is too long, 767 chars total.
Trying to shorten...
New name is archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com@kaggle-161607.iam.gserviceaccount.com%2F20210608%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210608T163726Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=hos.
--2021-06-08 16:37:53--  https://storage.googleapis.com/kaggle-data-sets/7415/10564/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210608%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210608T163726Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=9b8bb3b3e0daebd7132e73f4c4e73243e9c1747c81b5dcbb675268e3e196f11f0eb77a4d96ae1e2effa2ae0aaecf71ec4ffc30664dc08a2e4df54e51dc4f386010550912f80266e7872eb309c2d78242448f897a62d6667d3ff358f30c65db3538de8f6ef1b87d505ce1447ec74e93b1d7177272ae04f3e4f9bac9d7a73f72a1f12161056d3cd1bded96c05b74e481c5139a9361be7a7576e4bef3205d17c148e198f3d46637d813ac694d

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
!cp "/content/gdrive/MyDrive/archive.zip" "/content/archive.zip" 

In [ ]:
import shutil

In [ ]:
shutil

In [ ]:
import os

In [ ]:
os.listdir('/content')

['.config',
 '10288',
 '16896',
 '12867',
 '12868',
 '12822',
 '9258',
 '16550',
 '9383',
 '13458',
 '9176',
 '10254',
 '12748',
 '15902',
 '9135',
 '14192',
 '12879',
 '12824',
 '13106',
 '14155',
 '12870',
 '13617',
 '9323',
 '13687',
 '14153',
 '10258',
 '15472',
 '15632',
 '12906',
 '15515',
 '12818',
 '10307',
 '13666',
 '9124',
 '10274',
 '12884',
 '12898',
 '15514',
 '13462',
 '16554',
 '13020',
 '15473',
 '9043',
 '12948',
 '9291',
 '15512',
 '14306',
 '14191',
 '9228',
 '12752',
 '12877',
 '8959',
 '10291',
 '10277',
 '9125',
 '10295',
 '13460',
 '14213',
 '15634',
 '9265',
 '9266',
 '8975',
 '10264',
 '12242',
 '9381',
 '9324',
 '16552',
 '10299',
 '8951',
 '9382',
 '10292',
 '12932',
 '9081',
 '9260',
 '10290',
 '12954',
 '9227',
 '14082',
 '12894',
 '10269',
 '13613',
 '8864',
 '12930',
 '10255',
 '16531',
 '8974',
 '9029',
 '9319',
 '9126',
 '8980',
 '16570',
 '13024',
 '12886',
 '10272',
 '14321',
 '10300',
 '15513',
 '13403',
 '10302',
 '16167',
 '10276',
 '9041',
 '1287

In [ ]:
os.remove(i for i in ['.config',
 '10288',
 '16896',
 '12867',
 '12868',
 '12822',
 '9258',
 '16550',
 '9383',
 '13458',
 '9176',
 '10254',
 '12748',
 '15902',
 '9135',
 '14192',
 '12879',
 '12824',
 '13106',
 '14155',
 '12870',
 '13617',
 '9323',
 '13687',
 '14153',
 '10258',
 '15472',
 '15632',
 '12906',
 '15515',
 '12818',
 '10307',
 '13666',
 '9124',
 '10274',
 '12884',
 '12898',
 '15514',
 '13462',
 '16554',
 '13020',
 '15473',
 '9043',
 '12948',
 '9291',
 '15512',
 '14306',
 '14191',
 '9228',
 '12752',
 '12877',
 '8959',
 '10291',
 '10277',
 '9125',
 '10295',
 '13460',
 '14213',
 '15634',
 '9265',
 '9266',
 '8975',
 '10264',
 '12242',
 '9381',
 '9324',
 '16552',
 '10299',
 '8951',
 '9382',
 '10292',
 '12932',
 '9081',
 '9260',
 '10290',
 '12954',
 '9227',
 '14082',
 '12894',
 '10269',
 '13613',
 '8864',
 '12930',
 '10255',
 '16531',
 '8974',
 '9029',
 '9319',
 '9126',
 '8980',
 '16570',
 '13024',
 '12886',
 '10272',
 '14321',
 '10300',
 '15513',
 '13403',
 '10302',
 '16167',
 '10276',
 '9041',
 '12873',
 '14189',
 '12810',
 '14211',
 '10308',
 '13400',
 '10279',
 '16568',
 '13019',
 '12626',
 '13688',
 '13022',
 '12934',
 '13916',
 '12826',
 '8916',
 '16532',
 '12823',
 '10304',
 '12900',
 '9036',
 '9173',
 '12241',
 '13591',
 '9255',
 '12819',
 '10259',
 '13023',
 '14079',
 '8865',
 '9344',
 '9073',
 'hello.py',
 '15839',
 '14209',
 '13402',
 '10253',
 '10273',
 '10305',
 '12931',
 '16085',
 '10262',
 '13459',
 '10275',
 '15471',
 '10286',
 '8950',
 '13025',
 '16534',
 '12935',
 '14154',
 '8918',
 '14078',
 '14210',
 '12892',
 '9181',
 'archive.zip',
 '15840',
 '9262',
 '9023',
 '9123',
 '12750',
 '10257',
 '12947',
 '13689',
 '12908',
 '9075',
 '12910',
 '14156',
 '9259',
 '8956',
 '14212',
 '16555',
 '9321',
 '8957',
 '13694',
 '15903',
 '9257',
 '13401',
 '12897',
 '16165',
 '9174',
 '12929',
 '9076',
 '10285',
 '9037',
 'gdrive',
 '12869',
 '14081',
 '12749',
 '12820',
 '16533',
 '10278',
 '16895',
 '13018',
 '10301',
 '9254',
 '8867',
 '12893',
 '16551',
 '9320',
 '12821',
 '13461',
 '9347',
 '10261',
 '.ipynb_checkpoints',
 '12907',
 '12905',
 '10293',
 '9177',
 '14190',
 '16014',
 '10306',
 '9022',
 '8917',
 '10260',
 '13691',
 '12911',
 '9261',
 '9267',
 '9345',
 '9250',
 '9078',
 '13693',
 '14188',
 '10303',
 '9178',
 '12895',
 '12751',
 '13404',
 '10268',
 '10256',
 '12951',
 '12901',
 '12811',
 '12933',
 '13021',
 '8863',
 '12890',
 '8984',
 '8913',
 '14305',
 '14304',
 '12878',
 '12817',
 '12949',
 '15633',
 '12882',
 '13692',
 '9035',
 '9175',
 '12909',
 'IDC_regular_ps50_idx5',
 '12876',
 '12891',
 '16166',
 '9325',
 '12881',
 '9322',
 '12896',
 '8955',
 '15516',
 '12880',
 '14157',
 '12872',
 '10282',
 '13616',
 '9346',
 '16569')

TypeError: ignored

In [ ]:
!unzip -qq "/content/archive.zip" 

Streaming output truncated to the last 5000 lines.
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1551_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1601_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1651_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1701_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1751_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1801_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1851_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1901_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y1951_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y2001_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y2251_class0.png  
  inflating: IDC_regular_ps50_idx5/9346/0/9346_idx5_x2351_y2301_class0.png  
  inflating: IDC_regular_

In [ ]:
!cp "/content/Dataset.zip" "/content/gdrive/MyDrive/Dataset.zip"

In [ ]:
!cp "/content/archive.zip" "/content/gdrive/MyDrive/archive.zip"

In [ ]:
!wget "https://dl.dropboxusercontent.com/s/f01rtchlh9z9l5b/UCI%20HAR%20Dataset.zip"
!unzip "/content/UCI HAR Dataset.zip"

--2021-06-08 07:04:59--  https://dl.dropboxusercontent.com/s/f01rtchlh9z9l5b/UCI%20HAR%20Dataset.zip
Resolving dl.dropboxusercontent.com (dl.dropboxusercontent.com)... 162.125.2.15, 2620:100:6017:15::a27d:20f
Connecting to dl.dropboxusercontent.com (dl.dropboxusercontent.com)|162.125.2.15|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 60999314 (58M) [application/zip]
Saving to: ‘UCI HAR Dataset.zip’

UCI HAR Dataset.zip 100%[===================>]  58.17M   160MB/s    in 0.4s    

2021-06-08 07:05:00 (160 MB/s) - ‘UCI HAR Dataset.zip’ saved [60999314/60999314]

Archive:  /content/UCI HAR Dataset.zip
   creating: UCI HAR Dataset/
  inflating: UCI HAR Dataset/.DS_Store  
   creating: __MACOSX/
   creating: __MACOSX/UCI HAR Dataset/
  inflating: __MACOSX/UCI HAR Dataset/._.DS_Store  
  inflating: UCI HAR Dataset/activity_labels.txt  
  inflating: __MACOSX/UCI HAR Dataset/._activity_labels.txt  
  inflating: UCI HAR Dataset/features.txt  
  inflating: __MACOSX/UCI

In [ ]:
!pip install pycm==3.1
!sudo apt-get install tree

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  tree
0 upgraded, 1 newly installed, 0 to remove and 39 not upgraded.
Need to get 40.7 kB of archives.
After this operation, 105 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 tree amd64 1.7.0-5 [40.7 kB]
Fetched 40.7 kB in 1s (80.9 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package tree.
(Reading database ... 160772 files and directories currently instal

In [ ]:
import pandas as pd
import numpy as np
import os, glob
np.random.seed(21)
from pycm import ConfusionMatrix

In [ ]:
!tree "/content/UCI HAR Dataset"

/content/UCI HAR Dataset
├── activity_labels.txt
├── features_info.txt
├── features.txt
├── README.txt
├── test
│   ├── Inertial Signals
│   │   ├── body_acc_x_test.txt
│   │   ├── body_acc_y_test.txt
│   │   ├── body_acc_z_test.txt
│   │   ├── body_gyro_x_test.txt
│   │   ├── body_gyro_y_test.txt
│   │   ├── body_gyro_z_test.txt
│   │   ├── total_acc_x_test.txt
│   │   ├── total_acc_y_test.txt
│   │   └── total_acc_z_test.txt
│   ├── subject_test.txt
│   ├── X_test.txt
│   └── y_test.txt
└── train
    ├── Inertial Signals
    │   ├── body_acc_x_train.txt
    │   ├── body_acc_y_train.txt
    │   ├── body_acc_z_train.txt
    │   ├── body_gyro_x_train.txt
    │   ├── body_gyro_y_train.txt
    │   ├── body_gyro_z_train.txt
    │   ├── total_acc_x_train.txt
    │   ├── total_acc_y_train.txt
    │   └── total_acc_z_train.txt
    ├── subject_train.txt
    ├── X_train.txt
    └── y_train.txt

4 directories, 28 files


In [ ]:
#Features Description
!cat "/content/UCI HAR Dataset/features_info.txt" 

Feature Selection 

The features selected for this database come from the accelerometer and gyroscope 3-axial raw signals tAcc-XYZ and tGyro-XYZ. These time domain signals (prefix 't' to denote time) were captured at a constant rate of 50 Hz. Then they were filtered using a median filter and a 3rd order low pass Butterworth filter with a corner frequency of 20 Hz to remove noise. Similarly, the acceleration signal was then separated into body and gravity acceleration signals (tBodyAcc-XYZ and tGravityAcc-XYZ) using another low pass Butterworth filter with a corner frequency of 0.3 Hz. 

Subsequently, the body linear acceleration and angular velocity were derived in time to obtain Jerk signals (tBodyAccJerk-XYZ and tBodyGyroJerk-XYZ). Also the magnitude of these three-dimensional signals were calculated using the Euclidean norm (tBodyAccMag, tGravityAccMag, tBodyAccJerkMag, tBodyGyroMag, tBodyGyroJerkMag). 

Finally a Fast Fourier Transform (FFT) was applied to some of these signals pro

In [ ]:
#List of Features
!cat "/content/UCI HAR Dataset/features.txt" 

1 tBodyAcc-mean()-X
2 tBodyAcc-mean()-Y
3 tBodyAcc-mean()-Z
4 tBodyAcc-std()-X
5 tBodyAcc-std()-Y
6 tBodyAcc-std()-Z
7 tBodyAcc-mad()-X
8 tBodyAcc-mad()-Y
9 tBodyAcc-mad()-Z
10 tBodyAcc-max()-X
11 tBodyAcc-max()-Y
12 tBodyAcc-max()-Z
13 tBodyAcc-min()-X
14 tBodyAcc-min()-Y
15 tBodyAcc-min()-Z
16 tBodyAcc-sma()
17 tBodyAcc-energy()-X
18 tBodyAcc-energy()-Y
19 tBodyAcc-energy()-Z
20 tBodyAcc-iqr()-X
21 tBodyAcc-iqr()-Y
22 tBodyAcc-iqr()-Z
23 tBodyAcc-entropy()-X
24 tBodyAcc-entropy()-Y
25 tBodyAcc-entropy()-Z
26 tBodyAcc-arCoeff()-X,1
27 tBodyAcc-arCoeff()-X,2
28 tBodyAcc-arCoeff()-X,3
29 tBodyAcc-arCoeff()-X,4
30 tBodyAcc-arCoeff()-Y,1
31 tBodyAcc-arCoeff()-Y,2
32 tBodyAcc-arCoeff()-Y,3
33 tBodyAcc-arCoeff()-Y,4
34 tBodyAcc-arCoeff()-Z,1
35 tBodyAcc-arCoeff()-Z,2
36 tBodyAcc-arCoeff()-Z,3
37 tBodyAcc-arCoeff()-Z,4
38 tBodyAcc-correlation()-X,Y
39 tBodyAcc-correlation()-X,Z
40 tBodyAcc-correlation()-Y,Z
41 tGravityAcc-mean()-X
42 tGravityAcc-mean()-Y
43 tGravityAcc-mean()-Z
44 tGravityAc

In [ ]:
#Multiclass - Classification Labels
!cat "/content/UCI HAR Dataset/activity_labels.txt" 

1 WALKING
2 WALKING_UPSTAIRS
3 WALKING_DOWNSTAIRS
4 SITTING
5 STANDING
6 LAYING


In the given dataset, the experiments have been carried out with a group of 30 volunteers. They have captured 3-axial linear acceleration and 3-axial angular velocity using accelerometer and gyroscope. The **Raw signals** are found inside the **UCI HAR Dataset/train/Inertial Signals** folder. They have then labelled the data with **six activities** of person namely **WALKING, WALKINGUPSTAIRS, WALKINGDOWNSTAIRS, SITTING, STANDING, and  LAYING.**.

They have then pre-processed the data to extract various features like mean, standard deviation, Median absolute deviation, Largest value in array, Smallest value in array, Signal magnitude area,  Ener, Autorregresion coefficients with Burg order equal to 4, correlation coefficient between two signals, index of the frequency component with largest magnitude, Weighted average of the frequency components to obtain a mean frequency, 
skewness, kurtosis, bandsEnergy and angle between to vectors.

Finally they ended up with **561 Features** for accelerometer and gyroscope data. They then splitted the data into **70% Train and 30% Test**

The processed Train data can be found in the folder ** /content/UCI HAR Dataset/train/X_train.txt**

The coresponding training labels can be found in **/content/UCI HAR Dataset/train/y_train.txt **

The processed Test data can be found in the folder ** /content/UCI HAR Dataset/test/X_test.txt**

The coresponding Test labels can be found in **/content/UCI HAR Dataset/test/y_test.txt  **


------------------------------

**2. Apply atleast three ML algorithms and anlyze the results using**
**(a) confusion matrix.**
**(b) standard metrics.**

**Write your inference for each model.**
**Each algorithm and its analysis along with the inference for each of the used ML algorithm carries
15 x 3 = 45 marks).**

In [ ]:
df = pd.read_csv("/content/UCI HAR Dataset/train/X_train.txt", delim_whitespace=True, header=None);df.head()

0         1         2    ...       558       559       560
0  0.288585 -0.020294 -0.132905  ... -0.841247  0.179941 -0.058627
1  0.278419 -0.016411 -0.123520  ... -0.844788  0.180289 -0.054317
2  0.279653 -0.019467 -0.113462  ... -0.848933  0.180637 -0.049118
3  0.279174 -0.026201 -0.123283  ... -0.848649  0.181935 -0.047663
4  0.276629 -0.016570 -0.115362  ... -0.847865  0.185151 -0.043892

[5 rows x 561 columns]

In [ ]:
#Dataset Preparation

X_train = pd.read_csv('/content/UCI HAR Dataset/train/X_train.txt',delim_whitespace=True,header=None).to_numpy()
y_train = pd.read_csv('/content/UCI HAR Dataset/train/y_train.txt',delim_whitespace=True,header=None).to_numpy().reshape(-1)
X_test = pd.read_csv('/content/UCI HAR Dataset/test/X_test.txt',delim_whitespace=True,header=None).to_numpy()
y_test = pd.read_csv('/content/UCI HAR Dataset/test/y_test.txt',delim_whitespace=True,header=None).to_numpy().reshape(-1)

**Support Vector Machine (SVM)**




In [ ]:
from sklearn import svm
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [ ]:
#SVM with RBF Kernal
classifier = svm.SVC(gamma=0.001, kernel='rbf')
classifier.fit(X_train, y_train)
svm_predicted = classifier.predict(X_test)

print(classification_report(y_test, svm_predicted))
print("Confusion Matrix\n", ConfusionMatrix(y_test, svm_predicted))

              precision    recall  f1-score   support

           1       0.92      0.99      0.95       496
           2       0.93      0.94      0.94       471
           3       0.98      0.88      0.93       420
           4       0.90      0.83      0.86       491
           5       0.85      0.91      0.88       532
           6       1.00      1.00      1.00       537

    accuracy                           0.93      2947
   macro avg       0.93      0.93      0.93      2947
weighted avg       0.93      0.93      0.93      2947

Confusion Matrix
 Predict   1         2         3         4         5         6         
Actual
1         491       0         5         0         0         0         

2         24        445       2         0         0         0         

3         18        32        370       0         0         0         

4         0         2         0         406       83        0         

5         0         0         0         47        485       0         

6

In [ ]:
# SVM WIth Linear Kernal

classifier = svm.SVC(gamma=0.001, kernel='linear')
classifier.fit(X_train, y_train)
svm_predicted = classifier.predict(X_test)

print(classification_report(y_test, svm_predicted))
print("Confusion Matrix\n", ConfusionMatrix(y_test, svm_predicted))

              precision    recall  f1-score   support

           1       0.96      0.99      0.97       496
           2       0.98      0.96      0.97       471
           3       0.99      0.98      0.98       420
           4       0.96      0.89      0.92       491
           5       0.91      0.97      0.94       532
           6       1.00      1.00      1.00       537

    accuracy                           0.96      2947
   macro avg       0.97      0.96      0.96      2947
weighted avg       0.96      0.96      0.96      2947

Confusion Matrix
 Predict   1         2         3         4         5         6         
Actual
1         492       1         3         0         0         0         

2         18        451       2         0         0         0         

3         4         6         410       0         0         0         

4         0         2         0         435       54        0         

5         0         0         0         16        516       0         

6

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
  
# defining hyperparameter range
param_grid = {'gamma': [1, 0.01, 0.001],
              'kernel': ['rbf', 'linear']} 
  
grid = GridSearchCV(SVC(), param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train, y_train)


# print best parameter after tuning
print(grid.best_params_)
  
# print how our model looks after hyper-parameter tuning
print(grid.best_estimator_)

grid_predictions = grid.predict(X_test)
  
# print classification report
print(classification_report(y_test, grid_predictions))
print("Confusion Matrix\n", ConfusionMatrix(y_test, grid_predictions))

Fitting 5 folds for each of 6 candidates, totalling 30 fits
[CV] gamma=1, kernel=rbf .............................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ................. gamma=1, kernel=rbf, score=0.193, total=  59.5s
[CV] gamma=1, kernel=rbf .............................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:   59.5s remaining:    0.0s


[CV] ................. gamma=1, kernel=rbf, score=0.199, total=  59.5s
[CV] gamma=1, kernel=rbf .............................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:  2.0min remaining:    0.0s


[CV] ................. gamma=1, kernel=rbf, score=0.194, total=  59.7s
[CV] gamma=1, kernel=rbf .............................................
[CV] ................. gamma=1, kernel=rbf, score=0.196, total=  59.9s
[CV] gamma=1, kernel=rbf .............................................
[CV] ................. gamma=1, kernel=rbf, score=0.194, total=  59.8s
[CV] gamma=1, kernel=linear ..........................................
[CV] .............. gamma=1, kernel=linear, score=0.925, total=   2.5s
[CV] gamma=1, kernel=linear ..........................................
[CV] .............. gamma=1, kernel=linear, score=0.901, total=   2.6s
[CV] gamma=1, kernel=linear ..........................................
[CV] .............. gamma=1, kernel=linear, score=0.960, total=   2.7s
[CV] gamma=1, kernel=linear ..........................................
[CV] .............. gamma=1, kernel=linear, score=0.946, total=   2.8s
[CV] gamma=1, kernel=linear ..........................................
[CV] .

[Parallel(n_jobs=1)]: Done  30 out of  30 | elapsed:  7.2min finished


{'gamma': 1, 'kernel': 'linear'}
SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=1, kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)
              precision    recall  f1-score   support

           1       0.96      0.99      0.97       496
           2       0.98      0.96      0.97       471
           3       0.99      0.98      0.98       420
           4       0.96      0.89      0.92       491
           5       0.91      0.97      0.94       532
           6       1.00      1.00      1.00       537

    accuracy                           0.96      2947
   macro avg       0.97      0.96      0.96      2947
weighted avg       0.96      0.96      0.96      2947

Confusion Matrix
 Predict   1         2         3         4         5         6         
Actual
1         492       1         3         0         0         0         

2        

SVM Classifier was used to predict six Human activity classes.

RBF Kernal and Linear Kernals were used intially to test. We also perfomed a grid search to arrive at the best Hyperparammeter Gamma.

**RBF Kernal**: Accuracy -> **93%**

**Linear Kernal**: Accuracy -> **96%**


**From the results of the Grd Search, we observe that SVM with Linear Kernal with Gamma=1 was able to give a maximum accuracy of 96%.**

**Decision Tree Classifier**

In [ ]:
from sklearn import tree

Dtree = tree.DecisionTreeClassifier()
Dtree.fit(X_train, y_train)

Dtree_predicted = Dtree.predict(X_test)

print(classification_report(y_test, Dtree_predicted))
print("Confusion Matrix\n", ConfusionMatrix(y_test, Dtree_predicted))

#tree.plot_tree(Dtree) 

              precision    recall  f1-score   support

           1       0.83      0.88      0.86       496
           2       0.82      0.79      0.80       471
           3       0.88      0.86      0.87       420
           4       0.83      0.75      0.79       491
           5       0.79      0.86      0.82       532
           6       1.00      1.00      1.00       537

    accuracy                           0.86      2947
   macro avg       0.86      0.86      0.86      2947
weighted avg       0.86      0.86      0.86      2947

Confusion Matrix
 Predict   1         2         3         4         5         6         
Actual
1         437       36        23        0         0         0         

2         73        370       28        0         0         0         

3         15        45        360       0         0         0         

4         0         0         0         370       121       0         

5         0         0         0         76        456       0         

6

In [ ]:
dec_tree = tree.DecisionTreeClassifier()

# defining hyperparameter range
param_grid = {'max_depth': [1, 50, 100, 150, 250]} 
  
grid = GridSearchCV(dec_tree, param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train, y_train)


# print best parameter after tuning
print(grid.best_params_)
  
# print how our model looks after hyper-parameter tuning
print(grid.best_estimator_)

grid_predictions = grid.predict(X_test)
  
# print classification report
print(classification_report(y_test, grid_predictions))
print("Confusion Matrix\n", ConfusionMatrix(y_test, grid_predictions))

Fitting 5 folds for each of 5 candidates, totalling 25 fits
[CV] max_depth=1 .....................................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ......................... max_depth=1, score=0.379, total=   0.5s
[CV] max_depth=1 .....................................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.5s remaining:    0.0s


[CV] ......................... max_depth=1, score=0.377, total=   0.5s
[CV] max_depth=1 .....................................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.9s remaining:    0.0s


[CV] ......................... max_depth=1, score=0.378, total=   0.5s
[CV] max_depth=1 .....................................................
[CV] ......................... max_depth=1, score=0.378, total=   0.5s
[CV] max_depth=1 .....................................................
[CV] ......................... max_depth=1, score=0.378, total=   0.5s
[CV] max_depth=50 ....................................................
[CV] ........................ max_depth=50, score=0.802, total=   4.1s
[CV] max_depth=50 ....................................................
[CV] ........................ max_depth=50, score=0.810, total=   3.3s
[CV] max_depth=50 ....................................................
[CV] ........................ max_depth=50, score=0.827, total=   3.4s
[CV] max_depth=50 ....................................................
[CV] ........................ max_depth=50, score=0.886, total=   3.3s
[CV] max_depth=50 ....................................................
[CV] .

[Parallel(n_jobs=1)]: Done  25 out of  25 | elapsed:  1.2min finished


{'max_depth': 100}
DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=100, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')
              precision    recall  f1-score   support

           1       0.84      0.91      0.87       496
           2       0.84      0.80      0.82       471
           3       0.87      0.83      0.85       420
           4       0.83      0.77      0.80       491
           5       0.80      0.86      0.83       532
           6       1.00      1.00      1.00       537

    accuracy                           0.87      2947
   macro avg       0.86      0.86      0.86      2947
weighted avg       0.87      0.87      0.86      2947

Confusion Ma

Decision Tree Classifier was used to predict six Human activity classes.

We also perfomed a grid search to arrive at the best Hyperparameter Max_depth.

Accuracy -> **87%**

From the results of GridSearch, we observe that Decision Tree Classifier was able to give a maximum accuracy of **87%** with Hyper-Paramaeter **max_depth: 100**.

**Random Forest Classifier**

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier()
rf.fit(X_train, y_train)
rf_predicted = rf.predict(X_test)

print(classification_report(y_test, rf_predicted))
print("Confusion Matrix\n", ConfusionMatrix(y_test, rf_predicted))

              precision    recall  f1-score   support

           1       0.88      0.97      0.93       496
           2       0.90      0.90      0.90       471
           3       0.96      0.84      0.90       420
           4       0.90      0.88      0.89       491
           5       0.89      0.91      0.90       532
           6       1.00      1.00      1.00       537

    accuracy                           0.92      2947
   macro avg       0.92      0.92      0.92      2947
weighted avg       0.92      0.92      0.92      2947

Confusion Matrix
 Predict   1         2         3         4         5         6         
Actual
1         482       5         9         0         0         0         

2         40        425       6         0         0         0         

3         24        43        353       0         0         0         

4         0         0         0         433       58        0         

5         0         0         0         46        486       0         

6

In [ ]:
# defining hyperparameter range
param_grid = { 'max_depth': [5, 50, 100], 'n_estimators': [5, 50, 100, 150]}

grid = GridSearchCV(rf, param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train, y_train)


# print best parameter after tuning
print(grid.best_params_)
  
# print how our model looks after hyper-parameter tuning
print(grid.best_estimator_)

grid_predictions = grid.predict(X_test)
  
# print classification report
print(classification_report(y_test, grid_predictions))
print("Confusion Matrix\n", ConfusionMatrix(y_test, grid_predictions))

Fitting 5 folds for each of 12 candidates, totalling 60 fits
[CV] max_depth=5, n_estimators=5 .....................................


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV] ......... max_depth=5, n_estimators=5, score=0.848, total=   0.3s
[CV] max_depth=5, n_estimators=5 .....................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    0.3s remaining:    0.0s


[CV] ......... max_depth=5, n_estimators=5, score=0.856, total=   0.3s
[CV] max_depth=5, n_estimators=5 .....................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    0.6s remaining:    0.0s


[CV] ......... max_depth=5, n_estimators=5, score=0.884, total=   0.3s
[CV] max_depth=5, n_estimators=5 .....................................
[CV] ......... max_depth=5, n_estimators=5, score=0.885, total=   0.3s
[CV] max_depth=5, n_estimators=5 .....................................
[CV] ......... max_depth=5, n_estimators=5, score=0.912, total=   0.3s
[CV] max_depth=5, n_estimators=50 ....................................
[CV] ........ max_depth=5, n_estimators=50, score=0.877, total=   2.4s
[CV] max_depth=5, n_estimators=50 ....................................
[CV] ........ max_depth=5, n_estimators=50, score=0.903, total=   2.5s
[CV] max_depth=5, n_estimators=50 ....................................
[CV] ........ max_depth=5, n_estimators=50, score=0.910, total=   2.4s
[CV] max_depth=5, n_estimators=50 ....................................
[CV] ........ max_depth=5, n_estimators=50, score=0.899, total=   2.5s
[CV] max_depth=5, n_estimators=50 ....................................
[CV] .

[Parallel(n_jobs=1)]: Done  60 out of  60 | elapsed:  5.8min finished


{'max_depth': 100, 'n_estimators': 150}
RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=100, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=150,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)
              precision    recall  f1-score   support

           1       0.90      0.97      0.93       496
           2       0.90      0.92      0.91       471
           3       0.95      0.85      0.90       420
           4       0.90      0.88      0.89       491
           5       0.89      0.91      0.90       532
           6       1.00      1.00      1.00       537

    accuracy                      

Random Forest Classifier was used to predict six Human activity classes.

We also perfomed a grid search to arrive at the best Hyperparammeter Max_depth and n-estimator(Maximum number of trees).

Accuracy -> **92%**

From the results of GridSearch, we observe that Random Forest Classifier gave a maximum accracy of **92%** with **'max_depth': 100**, **'n_estimators': 150**

------------------------

**Overall Inference:**

Out of three ML Algorithms namely Support Vector Machines, Decision Trees and Random Forest, we observe that **SVM performed with higher accuracy of 96%** compared to others. We also observe that classwise Evaluation Metrics namely Precision, Recall, F1-Score were also higher in SVM Classifier.


---------------------

